pkg/controllers/user/globalnamespacerbac/rbac_common.go

```diff
@@ -0,0 +1,189 @@
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      roleName,
+					Namespace: namespace.GlobalNamespace,
+				},
+				Rules: []k8srbacv1.PolicyRule{
+					{
+						APIGroups:     []string{"management.cattle.io"},
+						Resources:     []string{resource},
+						ResourceNames: []string{resourceName},
+						Verbs:         verbs,
+					},
+				},
+			}
+			currentRole, err = userContext.Management.RBAC.Roles("").Create(role)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+	return currentRole, nil
+}
+
+func createRoleBinding(roleAccess string, name string, subjects []k8srbacv1.Subject, userContext *config.UserContext) (*k8srbacv1.RoleBinding, error) {
+	roleName, _ := getRoleNameAndVerbs(roleAccess, name)
+	// we can define the rolebinding first, since if it's not already present we can call create. And if it's present then we'll
+	// still need to compare the current members' list
+	newRoleBinding := &k8srbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleName,
+			Namespace: namespace.GlobalNamespace,
+		},
+		RoleRef: k8srbacv1.RoleRef{
+			Name: roleName,
+			Kind: "Role",
+		},
+		Subjects: subjects,
+	}
+	roleBinding, err := userContext.Management.RBAC.RoleBindings("").GetNamespaced(namespace.GlobalNamespace, name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			_, err = userContext.Management.RBAC.RoleBindings("").Create(newRoleBinding)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	} else if roleBinding != nil {
+		if !reflect.DeepEqual(roleBinding, newRoleBinding) {
+			toUpdate := newRoleBinding.DeepCopy()
+			updated, err := userContext.Management.RBAC.RoleBindings("").Update(toUpdate)
+			if err != nil && !apierrors.IsConflict(err) {
+				return updated, err
+			}
+		}
+	}
+	return newRoleBinding, nil
+}
+
+func getRoleNameAndVerbs(roleAccess string, resourceName string) (string, []string) {
+	var roleName string
+	var verbs []string
+	switch roleAccess {
+	case allAccess:
+		roleName = resourceName + "-a"
+		verbs = []string{"*"}
+	case updateAccess:
+		roleName = resourceName + "-u"
+		verbs = []string{"update", "get", "list", "watch"}
+	case readOnlyAccess:
+		roleName = resourceName + "-r"
+		verbs = []string{"get", "list", "watch"}
+	}
+	return roleName, verbs
+}
+
+func buildSubjectForMember(member v3.Member) (k8srbacv1.Subject, error) {
+	var name, kind string
+	if member.UserName != "" {
+		name = member.UserName
+		kind = "User"
+	}
+
+	if member.GroupPrincipalName != "" {
+		if name != "" {
+			return k8srbacv1.Subject{}, fmt.Errorf("member %v has both username and groupPrincipalId set", member)
+		}
+		name = member.GroupPrincipalName
+		kind = "Group"
+	}
+
+	if member.GroupName != "" {
+		if name != "" {
+			return k8srbacv1.Subject{}, fmt.Errorf("member %v has both username and groupname set", member)
```

Comment:
not familiar with this logic, is it possible for member to have both GroupPrincipalName and GroupName?

Reply:
@kinarashah we are going to keep GroupPrincipalName/UserPrincipalName only